### PR TITLE
=htc #20789 fix double push of requestPrepOut in HTTP blueprint

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/20789
Special situations, demand special handling.

Sadly I was unable to provide a reproducer of the Play behaviour as isolated Akka test :-(
I was testing with the actual Play project with SNAPSHOT built Akka

Reproducer is: https://github.com/wsargent/reproduce-akka-20642

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 1]
```diff
@@ -416,17 +417,22 @@ private[http] object HttpServerBluePrint {
             log.warning(
               """Sending 2xx response before end of request was received...
                 |Note that the connection will be closed after this response. Also, many clients will not read early responses!
-                |Consider waiting for the request end before dispatching this response!""".stripMargin)
+                |Consider only issuing this response after the request data has been completely read!""".stripMargin)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: While I was here wanted to cleanup the early response warning to be more clear what we want from people here

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 2]
```diff
@@ -416,17 +417,22 @@ private[http] object HttpServerBluePrint {
...
-          // request parser to complete
-          if (close && isEarlyResponse) pull(requestParsingIn)
+            () ⇒ {
+              if (close) {
+                complete(responseCtxOut)
+                // when the client closes the connection, we need to pull onc more time to get the
+                // request parser to complete (it will cause an incoming EntityStreamError)
+                if (requestStart.expect100Continue) pull(requestParsingIn)
+              } else pull(httpResponseIn)
+            }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: notice that the `pull` now happens from inside the handler, and is *conditional* on the situation

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 3]
```diff
@@ -416,17 +417,22 @@ private[http] object HttpServerBluePrint {
...
-          if (close) complete(responseCtxOut)
-          // when the client closes the connection, we need to pull onc more time to get the
-          // request parser to complete
-          if (close && isEarlyResponse) pull(requestParsingIn)
+            () ⇒ {
+              if (close) {
+                complete(responseCtxOut)
+                // when the client closes the connection, we need to pull onc more time to get the
+                // request parser to complete (it will cause an incoming EntityStreamError)
+                if (requestStart.expect100Continue) pull(requestParsingIn)
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: this changes the original condition from `close && isEarlyResponse` to `close && expect100Continue`. Is this intentional? What is the reason for the change? (The other change is that these are executed only after `emit` finishes)

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: Thanks checking, I went back to see which of those two is really needed (had multiple phases of digging into this and that was one that worked).
So we don't need to delay the pull actually, all tests and the play reproducer work fine with the switch of the pull condition.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: I pushed a new commit with that

**Comment by ktoso** [comment]
PS: The line `if (close && isEarlyResponse) pull(requestParsingIn)` was introduced only recently as fixing of other things - seems it was not quite correct.
I'm not happy I was not able to reproduce hm :<

**Comment by typesafehub-validator** [comment]
Hi @drprofesq,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
lol, I used <nobody@example.com> as placeholder for author as I cherrypicked a gist by @Joe-Edwards not knowing his email, so typesafe validator asked the linked github account to it for CLA. 

Sorry for the noise @drprofesq, please ignore

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala (hunk 1 above)
Comment (patriknw) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: Log messages with line breaks are not nice, in my experience. Do we have a lot of those in http?

File: akka-http-core/src/test/scala/akka/http/scaladsl/ServerControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import akka.stream.scaladsl._
+import akka.stream.testkit.scaladsl.{ TestSink, TestSource }
+import akka.testkit.AkkaSpec
+import akka.util.ByteString
+
+class ServerControllerSpec extends AkkaSpec {
+
+  "HTTP Server Controller" should {
+    "handle early responses (Issue: #20642, #20789)" in {
+      implicit val system = ActorSystem()
```
Comment (patriknw) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ServerControllerSpec.scala]: when is this system terminated?

**Comment by patriknw** [comment]
I'm not sure I understand what you cherry-picked. Has Joe-Edwards signed the CLA if you used his code?

**Comment by ktoso** [comment]
The test was picked from @Joe-Edwards's snippet. 
I checked, he has not signed the CLA, so I'll remove the test from this PR - thanks for reminding me about this.

@Joe-Edwards if you'd like to contribute this commit: https://github.com/akka/akka/pull/20841/commits/c0837f63933313f57f7cc348f90d2c15955815d1 that would be very nice

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala (hunk 1 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: I don't think there's many of them, I can remove the multiline one here while I'm here

**Comment by ktoso** [comment]
Discussed this one a bit offline, will merge after validation
